polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -90,36 +90,243 @@ static bool isSmallerThanNegativeOne(Value val, DataFlowSolver &solver) {
   return (optConstVal.has_value() && optConstVal->slt(-1));
 }
 
-/// Remove conversion like operations in the incoming expression if any.
-static Value removeConversions(Value expr) {
-  if (isa<CastOpInterface>(expr.getDefiningOp()))
-    return expr.getDefiningOp()->getOperand(0);
-  return expr;
+/// Walks up the parents and records the ones with the specified type.
+template <typename T> static SetVector<T> getParentsOfType(Block *block) {
+  SetVector<T> res;
+  auto *current = block->getParentOp();
+  while (current) {
+    if (auto typedParent = dyn_cast<T>(current)) {
+      assert(res.count(typedParent) == 0 && "Already inserted");
+      res.insert(typedParent);
+    }
+    current = current->getParentOp();
+  }
+  return res;
 }
 
-/// Determine whether 'expr' is a multiplication involving 'val'.
-/// TODO: Rewrite this function to handle more generic expressions.
-/// Example:
-///   %0 = arith.index_cast %ii : index to i64
-///   %i = arith.muli %0, %c1 : i64
-/// In this case we fail to match because the %0 is not %ii.
-static std::optional<Value> getMultiplier(Value expr, Value val) {
-  expr = removeConversions(expr);
-  if (expr == val) {
-    OpBuilder b(expr.getContext());
-    return b.create<arith::ConstantIndexOp>(b.getUnknownLoc(), 1);
+namespace {
+
+/// Represents a multiplication factor.
+class Multiplier {
+  friend raw_ostream &operator<<(raw_ostream &, const Multiplier &);
+
+public:
+  Multiplier(Value val) : val(val) {}
+
+  bool operator==(const Multiplier &other) { return val == other.val; }
+
+  Value getValue() const { return val; }
+
+  /// Determine whether the multiplier has value one.
+  bool isOne(DataFlowSolver &solver) const {
+    if (auto constVal = dyn_cast<arith::ConstantIndexOp>(val.getDefiningOp())) {
+      if (constVal.value() == 1)
+        return true;
+    }
+    std::optional<APInt> constVal = getConstIntegerValue(val, solver);
+    return (constVal.has_value() && constVal->isOne());
```

Comment:
```suggestion
    std::optional<APInt> constVal = getConstIntegerValue(val, solver);
    return (constVal.has_value() && constVal->isOne());
```
Question: Do we need the first part?